File: src/main/java/dev/slimevr/vr/trackers/udp/UDPDevice.java
package dev.slimevr.vr.trackers.udp;

import dev.slimevr.NetworkProtocol;
import dev.slimevr.vr.Device;
import dev.slimevr.vr.trackers.IMUTracker;
import dev.slimevr.vr.trackers.Tracker;
import io.eiren.util.collections.FastList;

import java.net.InetAddress;
import java.net.SocketAddress;
import java.util.concurrent.ConcurrentHashMap;


public class UDPDevice extends Device {

	public final int id;
	public SocketAddress address;
	public InetAddress ipAddress;
	public long lastPacket = System.currentTimeMillis();
	public int lastPingPacketId = -1;
	public long lastPingPacketTime = 0;
	public String name;
	public String descriptiveName;
	public StringBuilder serialBuffer = new StringBuilder();
	public long lastSerialUpdate = 0;
	public long lastPacketNumber = -1;
	public NetworkProtocol protocol = null;
	public int firmwareBuild = 0;
	public boolean timedOut = false;
	private final FastList<Tracker> trackers = new FastList<>();
	private final ConcurrentHashMap<Integer, Integer> remote_id_map = new ConcurrentHashMap<>();

	public UDPDevice(SocketAddress address, InetAddress ipAddress) {
		this.address = address;
		this.ipAddress = ipAddress;
		this.id = UDPDevice.nextLocalDeviceId.incrementAndGet();
	}

	public boolean isNextPacket(long packetId) {
		if (packetId != 0 && packetId <= lastPacketNumber)
			return false;
		lastPacketNumber = packetId;
		return true;
	}

	@Override
	public String toString() {
		return "udp:/" + ipAddress;
	}

	@Override
	public int getId() {
		return id;
	}

	@Override
	public String getManufacturer() {
		return "SlimeVR";
	}

	@Override
	public String getDisplayName() {
		return name;
	}

	@Override
	public String getFirmwareVersion() {
		return "v" + firmwareBuild;
	}

	@Override
	public String getCustomName() {
		return name;
	}

	@Override
	public InetAddress getIpAddress() {
		return this.ipAddress;
	}

	@Override
	public FastList<Tracker> getTrackers() {
		return this.trackers;
	}

	public IMUTracker setupTrackerByRemoteId(int id) {
		if (remote_id_map.get(id) == null) {
			// New tracker id sent, create a mapping to the future
			// position of the tracker in the trackers list
			remote_id_map.put(id, this.trackers.size());
			return null; // There's no tracker set up yet
		}
		return getTrackerByRemoteId(id);
	}

	public IMUTracker getTrackerByRemoteId(int id) {
		Integer remote_id = remote_id_map.get(id);
		if (remote_id != null && remote_id < this.getTrackers().size())
			return (IMUTracker) this.getTrackers().get(remote_id);
		return null;
	}
}

File: server/core/src/main/java/dev/slimevr/tracking/trackers/udp/UDPDevice.kt
package dev.slimevr.tracking.trackers.udp

import dev.slimevr.NetworkProtocol
import dev.slimevr.VRServer
import dev.slimevr.tracking.trackers.Device
import dev.slimevr.tracking.trackers.Tracker
import java.net.InetAddress
import java.net.SocketAddress
import java.util.concurrent.ConcurrentHashMap
import kotlin.time.TimeSource

class UDPDevice(
	var address: SocketAddress,
	var ipAddress: InetAddress,
	override val hardwareIdentifier: String,
	override val boardType: BoardType = BoardType.UNKNOWN,
	override val mcuType: MCUType = MCUType.UNKNOWN,
	val timeslot: Int,
) : Device(true) {

	override val id: Int = nextLocalDeviceId.incrementAndGet()

	@JvmField
	var lastPacket = System.currentTimeMillis()

	@JvmField
	var lastPingPacketId = -1

	@JvmField
	var lastPingPacketTime: Long = 0
	var lastPingPacketInstant = TimeSource.Monotonic.markNow()

	override var name: String? = null
		set(name) {
			super.name = name
			field = name
		}

	@JvmField
	var descriptiveName: String? = null

	@JvmField
	var serialBuffer = StringBuilder()

	@JvmField
	var lastSerialUpdate: Long = 0

	@JvmField
	var lastPacketNumber: Long = -1

	@JvmField
	var totalPacketsReceived: Int = 0

	@JvmField
	var acceptedPackets: Int = 0

	@JvmField
	var lastPacketCounterReset: Long = System.currentTimeMillis()

	val packetLossPercent: Float
		get() = if (totalPacketsReceived == 0) 0f else (1f - acceptedPackets.toFloat() / totalPacketsReceived.toFloat())

	@JvmField
	var protocol: NetworkProtocol? = null

	@JvmField
	var protocolVersion = 0

	@JvmField
	var timedOut = false
	override val trackers = ConcurrentHashMap<Int, Tracker>()

	override suspend fun setMag(state: Boolean, sensorId: Int) {
		if (sensorId == 255) {
			VRServer.instance.trackersServer.setConfigFlag(this, ConfigTypeId(1u), state)
			trackers.forEach { (_, t) -> t.setMagPrivate(state) }
		} else {
			require(trackers[sensorId] != null) { "There is no tracker $sensorId in device ${toString()}" }
			VRServer.instance.trackersServer.setConfigFlag(this, ConfigTypeId(1u), state, sensorId)
			trackers[sensorId]!!.setMagPrivate(state)
		}
	}

	var firmwareFeatures = FirmwareFeatures()

	fun isNextPacket(packetId: Long): Boolean {
		val now = System.currentTimeMillis()
		if (now - lastPacketCounterReset >= 10_000L) {
			totalPacketsReceived = 0
			acceptedPackets = 0
			lastPacketCounterReset = now
		}
		totalPacketsReceived++
		val accepted = packetId == 0L || packetId > lastPacketNumber
		if (accepted) {
			lastPacketNumber = packetId
			acceptedPackets++
		}
		val lost = totalPacketsReceived - acceptedPackets
		trackers.values.forEach {
			it.packetsReceived = totalPacketsReceived
			it.packetsLost = lost
			it.packetLoss = packetLossPercent
		}
		return accepted
	}

	override fun toString(): String = "udp:/$ipAddress"

	override var manufacturer: String?
		get() = "SlimeVR"
		set(manufacturer) {
			super.manufacturer = manufacturer
		}

	fun getTracker(id: Int): Tracker? = trackers[id]
}
